File: 2227-sum-of-subarray-ranges/sum-of-subarray-ranges.cpp
class Solution {
public:
    vector<int> findNextSmallerElement(vector<int> &arr) {
        int n = arr.size();
        vector<int> nse(n);  
        stack<int> st;       

        for (int i = n - 1; i >= 0; i--) {
            while (!st.empty() && arr[st.top()] >= arr[i]) {
                st.pop();
            }
            nse[i] = st.empty() ? n : st.top();
            st.push(i);
        }
        return nse;
    }

    vector<int> findPreviousSmallerElement(vector<int> &arr) {
        int n = arr.size();
        vector<int> psee(n);  
        stack<int> st;       

        for (int i = 0; i < n; i++) {
            while (!st.empty() && arr[st.top()] > arr[i]) {
                st.pop();
            }
            psee[i] = st.empty() ? -1 : st.top();
            st.push(i);
        }
        return psee;
    }

    vector<int> findNextGreaterElement(vector<int> &arr) {
        int n = arr.size();
        vector<int> nge(n);  
        stack<int> st;       

        for (int i = n - 1; i >= 0; i--) {
            while (!st.empty() && arr[st.top()] <= arr[i]) {
                st.pop();
            }
            nge[i] = st.empty() ? n : st.top();
            st.push(i);
        }
        return nge;
    }

    vector<int> findPreviousGreaterElement(vector<int> &arr) {
        int n = arr.size();
        vector<int> pge(n);  
        stack<int> st;       

        for (int i = 0; i < n; i++) {
            while (!st.empty() && arr[st.top()] < arr[i]) {
                st.pop();
            }
            pge[i] = st.empty() ? -1 : st.top();
            st.push(i);
        }
        return pge;
    }

    long long sumSubarrayMins(vector<int>& arr) {
        int n = arr.size();
        vector<int> nse = findNextSmallerElement(arr);
        vector<int> psee = findPreviousSmallerElement(arr);
        long long total = 0;  

        for (int i = 0; i < n; i++) {
            long long left = i - psee[i];
            long long right = nse[i] - i;
            total += left * right * arr[i];
        }
        return total;
    }

    long long sumSubarrayMax(vector<int>& arr) {
        int n = arr.size();
        vector<int> nge = findNextGreaterElement(arr);
        vector<int> pge = findPreviousGreaterElement(arr);
        long long total = 0;  

        for (int i = 0; i < n; i++) {
            long long left = i - pge[i];
            long long right = nge[i] - i;
            total += left * right * arr[i];
        }
        return total;
    }

    long long subArrayRanges(vector<int>& nums) {
        return sumSubarrayMax(nums) - sumSubarrayMins(nums);
    }
};

File: 2227-sum-of-subarray-ranges/sum-of-subarray-ranges.java
class Solution {
    public long subArrayRanges(int[] nums) {
        return sumOfSubArrayMaximum(nums) - sumOfSubArrayMinimum(nums);
    }
    public long sumOfSubArrayMaximum(int[] nums){
        int[] nge = nextGreaterNumber(nums);
        int[] pge = previousGreaterNumber(nums);
        long result=0;
        for(int i=0;i<nums.length;i++){
            long leftIndex = i-pge[i];
            long rightIndex = nge[i]-i;
            result = result + (leftIndex * rightIndex * nums[i] * 1L);
        }
        return result;
    }
    public int[] nextGreaterNumber(int[] nums){
        int n = nums.length;
        int[] result = new int[n];
        Stack<Integer> stack = new Stack<>();
        for(int i=n-1;i>=0;i--){
            //explain
            while(!stack.isEmpty() && nums[stack.peek()]<=nums[i]){
                stack.pop();
            }
            result[i]=stack.isEmpty()?n:stack.peek();
            stack.push(i);
        }
        return result;
    }
    public int[] previousGreaterNumber(int[] nums){
        int n = nums.length;
        Stack<Integer> stack = new Stack<>();
        int[] result = new int[n];
        for(int i=0;i<n;i++){
            // explain
            while(!stack.isEmpty() && nums[stack.peek()]<nums[i]){
                stack.pop();
            }
            result[i]=stack.isEmpty()?-1:stack.peek();
            stack.push(i);
        }
        return result;
    }
    public long sumOfSubArrayMinimum(int[] nums){
        int[] pse = previousSmallerNumber(nums);
        int[] nse = nextSmallerNumber(nums);
        long result=0;
        for(int i=0;i<nums.length;i++){
            long leftIndex = i - pse[i];
            long rightIndex = nse[i]-i;
            result = result+(leftIndex * rightIndex * nums[i] *1L);
        }
        return result;
    }
    public int[] previousSmallerNumber(int[] nums){
        int n=nums.length;
        int[] result = new int[n];
        Stack<Integer> stack = new Stack<>();
        for(int i=0;i<n;i++){
            //explain
            while(!stack.isEmpty() && nums[stack.peek()]>nums[i]){
                stack.pop();
            }
            if(stack.isEmpty()){
                result[i]=-1;
            }
            else{
                result[i]=stack.peek();
            }
            stack.push(i);
        }
        return result;
    }
    public int[] nextSmallerNumber(int[] nums){
        int n= nums.length;
        int[] result = new int[n];
        Stack<Integer> stack = new Stack<>();
        for(int i=n-1;i>=0;i--){
            //explain
            while(!stack.isEmpty() && nums[stack.peek()]>=nums[i]){
                stack.pop();
            }
            result[i]=stack.isEmpty()?n:stack.peek();
            stack.push(i);
        }
        return result;
    }
}